Added functionality to locate elements based on their tagname, attribute, and value.

hamzaaitbrik opened the pull request:
Hello, I hope you're doing fine.

I submitted an issue not along ago stating that I had issues trying to locate elements nested deeply into shadowRoots. I tried doing it with JavaScript, the same way I do it with Selenium, but that didn't work because the `evaluate()` method only returns serializable data, and I needed elements to interact with.

So, I decided to change the source code of the project to fix my issue, I ended up creating 4 other functions that can accurately, and efficiently locate elements in the whole DOM, including shadowRoots and iFrames, using only the `tagname` of the element, and a pair of attribute/value to narrow down the targeting.

The functions I created are `locate()`, `locate_all()`, `locate_element_by_tagname_attribute_value()`, and `locate_elements_by_tagname_attribute_value()`.

The `locate()` function takes three arguments, `tagname`, `attribute`, and `value`. That way, if we're aiming to locate a button, and we know that our target button is the only button with a certain pair of `attribute`/`value`, we can effectively locate that button every single time, even if the DOM changes in case we're working on a website that changes its structure.

The `locate_all()` function takes three arguments as well, `tagname`, `attribute`, and `value`. It works the exact same way as the `locate()` function, the only difference is that we're able to retrieve a list of all elements of our specified `tagname`, and pair of `attribute`/`value`.

Feel free to examine the code I added, and let me know if I missed anything. I will be more than happy to modify on it to match the preferences of this project.

I thought of deleting the `find()` and `select()` functions as they won't be much of use compared to `locate()`, but that's your decision to make.

Looking forward for this pull request to get merged. I believe this will add great value!

Thanks,
Hamza

stephanlensky commented:
Thank you for your contribution!

I'm interested in merging this, but am hesitant because adding another "find" function with a new name when `find` and `select` already exist sounds like it might be confusing for new users.

I spent a bit of time thinking how to solve this problem, but wasn't able to come up with anything. I'd be open to suggestions, though. I do think it's important we don't add a 3rd new function with ambiguous purpose compared to the other existing ones.

hamzaaitbrik commented:
Thanks for the reply @stephanlensky.

I don't know what you suggest. I believe the function I added is much more efficient at locating elements than the other two functions, **especially** if elements are nested deeply into shadowRoots, hence why I thought of making such thing.

The find() and select() functions are **not able** to locate elements within shadowRoots, which I stumble upon quite frequently when doing web automation.

If having three functions is not an option, I would suggest removing the `find()` function altogether, because that's the least important one of them. Selecting elements based on their text can be achieved using the `locate()` function with the arguments `tagname=tagname`, `attribute=innerHTML` or `attribute=textContent`, and `value=ourTarget`.

The only downside in only using the `select()` and `locate()` functions solely is having to inspect the webpage whenever we're looking for our targeted element. I believe that should be totally okay for people who are using this project, because they're developers. I myself do web automation, and I can't imagine myself writing a script for a webpage without inspecting it's DOM.

Let me know what you think about what I said, and whether there's any modifications I should add.

I am also looking forward into having this pull request merged into the main branch.

Cheers,
Hamza

stephanlensky commented:
@hamzaaitbrik after giving this more thought, I think the best path is to integrate your functionality into the existing `find`/`find_all` functions, rather than creating a new function.

If you are familiar with the Python package [beautifulsoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/), my goal is for `find`/`find_all` to be very similar to the `find`/`find_all` functions from there.

More concretely, here is what I suggest:

1. Modify the existing `find`/`find_all` functions to traverse shadow roots
2. Add a new **optional** argument (we don't want to break existing users) to `find`/`find_all`: `attrs`.
    - This argument will be a dict mapping attribute name -> value
    - You can specify one or more attributes in this dict to filter results
    - Very similar to what you have now, but extended to work with multiple attributes instead of just one
3. Make the existing `text` and `best_match` arguments optional, so that you have the option to filter only by `attrs` without searching for some specific text.

Let me know if you have any questions or if this makes sense, thanks.

hamzaaitbrik commented:
Hello @stephanlensky,

I have given what you said so much thought. But I can't comprehend how would such function work? The `locate()` function takes a tagname as argument, that way we know what we're targeting; and that's more technical than just trying to find something called 'Log In' or 'Submit,' which is not practical in my opinion.

I genuinely believe that this project should ditch the option of finding elements by text, because that's simply not as efficient as finding elements by tagname and attribute/value.

I would argue that whoever will base their work on this project will definitely need a bulletproof method to locate elements effectively, and text simply doesn't do it. I tried, tho; I tried looking for buttons with 'Log In' text with a modified `find()` method, but I don't accurately get my target element every time, I sometimes get `div` or `span` elements while I need a button to click.

This will obviously be a big merge, adding/changing functions is something users will have to adapt with.
What I suggest is, we ditch finding elements by text altogether, because simply that is not efficient. I will be more than happy to rewrite the documentation for the project explaining how the function responsible for finding elements using their tagname and attrs work.

I'm open to other suggestions, but finding elements with text just doesn't work, I would consider myself intermediate in web automation and I know that finding the exact same element every single time is a crucial for longevity.

Let me know what you think.

stephanlensky commented:
@hamzaaitbrik please reread what I said carefully. Specifically:

> Make the existing text and best_match arguments optional, so that you have the option to filter only by attrs without searching for some specific text.

I am suggesting the existing function be extended in such a way that searching by text is no longer required, and you can instead search by attribute and tag name.

Attributes/attributes values can be provided as an optional dictionary. Similarly, tag name can be provided as an optional string.

This gives users the power to search however they want, whether that is by text, attribute or tag name, all without breaking any existing usages or adding a new function.

hamzaaitbrik commented:
Oh, okay! Gotcha now. Sorry, I totally missed the part where you said text could be optional. I will code it and try to make it as efficient as possible and submit the modified code. Thanks.

hamzaaitbrik commented:
Hello @stephanlensky, I hope you're doing fine. Sorry for the late reply, I was hiking the mountains.

Anyway, I was able to make more modifications towards the library to get something close to what you said.
Here are the changes I made:

- I modified the find and find_all functions so that users can search either by using tagname, attrs, or text; users can also search by tagname and attrs together. I decided not to merge the search functionality by text and tagname/attrs, users can only pick one, meaning: if a user chose to fill text, tagname, and attrs in the find or find_all functions; searching by tagname/attrs is gonna be favored, searching by text will only happen if the user inserts only text without specifying tagname and attrs.
- I added two functions, find_element_by_tagname_attrs and find_elements_by_tagname_attrs, and left find_element_by_text and find_elements_by_text unchanged. find_element_by_tagname_attrs and find_elements_by_tagname_attrs both takes two optional arguments, tagname and attrs; allowing users to search either by one of them, or preferably both.
- Only find_element_by_tagname_attrs and find_elements_by_tagname_attrs can traverse through shadowRoots, like I said; I left the functionality to find elements by text unchanged.
- Added functionality to wait for elements using either their tagname, attributes, or both.

I tested all functionality of the new code, but feel free to do so yourself.

Note that I will be more than happy to change the documentation of this library to explain the new changes.

Let me know if there is any more modifications needing to be added. Looking forward to having this merged.
Thanks,
Hamza

stephanlensky commented:
Hi @hamzaaitbrik,

Thanks for the update, this is looking much better now!

- Adding tagname, and attrs to the existing `find`/`find_all` is exactly what I had in mind, however I am not sure there is a reason not to allow searching both by text and tagname/attrs. I think users should be able to provide any combination of these parameters and have it work as expected. Would that be possible?
- To that end, I think it's okay if we update the text search to allow traversing shadowroots.
- Could you please format and lint the code? You can run the script in `scripts/format.sh` first to format, then run linting with `scripts/lint.sh`.

I apologize for asking for more from you, I'd like to get this merged as well. I'd just like to make sure we keep the interface intuitive so that new developers can easily pick up the library. Let me know your thoughts, thank you 🙂

hamzaaitbrik commented:
Hey man, no worries. I understand your need to imperfect this project for new developers to adopt it.

Honestly, I do not know how to implement the search by text, tagname, and attrs altogether. I thought of searching by tagname and attrs and narrowing it down with text..what do you think about that? The main reason why I chose to separate finding elements by text and tagname/attrs is because the latter is more efficient. I mean, if developers wanted to pinpoint a specific button, they can already narrow it down with multiple attributes...or, and I literally just had this pop into my mind haha, I could maybe insert the text value into the attrs dictionary as "innerText":"textValue"

I'll add that, then add the functionality to traverse through the DOM even when only text is provided...or I might end up removing the find_element_by_text and find_elements_by_text and use the {"innerText":"textValue"} as attrs. Let me know what you think.

I will make the changes I stated above, I will most likely end up transitioning the text functionality into {"innerText":"textValue"} and then I'm gonna format and lint the text, then I'm gonna let you know.

Cheers,
Hamza

stephanlensky commented:
That's alright, I agree it's a bit more challenging to do things this way.

For the implementation, I think we don't want to force users to specify `text` as part of `attrs`, but if you keep the function signature the same and then e.g. on the first line do like `attrs["innerText"] = text`, maybe it's okay.

If you aren't sure how to proceed, I'm happy to take a look as well and try to help. I should have some time in the next few weeks, but it might not happen today or tomorrow.

stephanlensky commented:
BTW, I do think we need to keep `find_elements_by_text` because it's part of the public API, so removing it would be a breaking change.

However, if you implement the functionality in `find`, you can just replace it with something like

```python
def find_elements_by_text(self, text, best_match):
    find(text=text, best_match=best_match)
```

I'm typing from my phone so this example isn't complete, but something like that.

hamzaaitbrik commented:
No worries man, I got this. I'll get back to you in a few days.

hamzaaitbrik commented:
Hello @stephanlensky.

I've done some changes to the repository, and done quite a lot of testing; and I concluded that searching by a combination of tagname, attrs, and text altogether is not efficient. It just does not return my targeted element every time.

I fell into a few problems, one of which is that the CDP object does not return an element and therefore I cannot directly check for its text content. I'm working with very complex JavaScript-rendered web pages, and sometimes my targeted element is an input tag that has a bunch of span/div elements inside of it which themselves hold the text value...I thought of searching by tagname and attrs, then narrowing down by text; but that just does not work because most of the times, on complex web pages, my targeted element holds no text, but instead, elements inside of it do. And that creates a mess while trying to narrow down by text.

I believe that the functionality I currently have is more than sufficient, allowing users to only search by text alone, or tagname/attrs on the other hand, but never both. I believe any beginner who is new to web automation can inspect web pages and collect the necessary data(tagname/attrs) to successfully target a web element.

Otherwise, I'd appreciate it if you look into the code and try to come up with a solution to search by all three arguments with no issue.

Cheers,
Hamza

hamzaaitbrik commented:
Look at this web page for example.
![image](https://github.com/user-attachments/assets/7fc5d72d-eecc-4a50-bf59-71ef18f5e04a)

As you can see, I am targeting an input tag that has the attribute name equals to username. While if I wanted to narrow down the search by text, people would usually aim for Email or username, but clearly that is a separate element that is not even a child of the input tag and has no relation to it.

I hope you consider that before making any decision. I personally believe that searching by tagname, attrs and text altogether is not efficient.

Let me know what you think.

stephanlensky commented:
Hi @hamzaaitbrik, thanks for making an effort here. I agree that searching by tag name/attribute is usually better (like in your example), but there are some cases where it's just not possible. For example, many webpages use auto-generated or even randomized IDs/classnames for elements, which can make it difficult to narrow down results. Sometimes, these elements will have other identifying attributes which you can search for, but now always. For those cases (and also to not break existing users of the library), we need to continue supporting search by text.

I do still believe it's possible to accomplish what I suggested, but I understand the difficulty here, especially because a different CDP API needs to be used when searching by text. I do not want to make perfection the enemy of progress, so here is what I recommend as a compromise:

1. Rather than integrate your changes with `find`/`find_all` like I originally suggested, let's scope it down a bit and only include the new `find_element_by_tagname_attrs`/`find_elements_by_tagname_attrs` functions with no text search support.
2. Revert your changes to `find`/`find_all`, maybe I or someone else can revisit that at a later time.
3. Format and lint the code with `./scripts/format.sh` and `./scripts/lint.sh`.

After you do that, I will give this a proper line-by-line review so that we can work towards merging it.

Thanks,
Stephan

stephanlensky commented:
> After you do that, I will give this a proper line-by-line review so that we can work towards merging it.

BTW re: this, I have not looked at the implementation in-depth yet, but just to let you know one of my comments is definitely going to be to ask that we share code between `find_element_by_tagname_attrs`/`find_elements_by_tagname_attrs` instead of duplicating the logic twice. We can save that time if you are able to address that now before the final review.

hamzaaitbrik commented:
Hello @stephanlensky, hope you're doing fine.

I managed to pull it off, now `find()` and `find_all()` can work with any combination of `tagname`, `attrs`, and `text`.
This is what I did:
- I changed the function `find_element_by_tagname_attrs` and renamed it to `find_element_by_tagname_attrs_text` so that whenever a node is retrieved, it makes an element with that node, and checks for `tagname`, `attrs`, `text` against that element's `tag_name`, `attributes`, and `text`.
- The same logic was applies to `find_elements_by_tagname_attrs_text`, I couldn't share the code between `find_element_by_tagname_attrs_text `and `find_elements_by_tagname_attrs_text` since the first one is always bound to return the first matching result.
- I changed `find_element_by_text` and `find_elements_by_text` so that they would call `find` and `find_all`.
- I changed `wait_for` and `__call__` to work with the new functionality

This is how an element is matched against the provided tagname, attrs, and text:
```
# create an element to check for the conditions we're looking for
elem = element.create(node, self, parent_tree)

# check for conditions
matches_tagname = (
    not tagname
    or (
        elem.tag_name
        and tagname.strip().lower() == elem.tag_name.strip().lower()
    )
)  # this condition evaluates to True if tagname was not provided; no filtering by tagname. Or if tagname equals our targeted element's tagname

matches_attrs = (
    not attrs
    or (
        elem.attributes
        and all(
            any(
                elem.attributes[i] == attr
                and value in elem.attributes[i + 1].split()
                for i in range(0, len(elem.attributes), 2)
            )
            for attr, value in attrs.items()
        )
    )
)  # this condition evaluates to True if attrs was not provided; no filtering by attrs. Or if the provided attrs are in our targeted element's attributes

matches_text = (
    not text
    or (elem.text and text.strip().lower() in elem.text.strip().lower())
)  # this condition evaluates to True if text was not provided; no filtering by text. Or if text is in our targeted element's text

# if all conditions match, we return the target element
if matches_tagname and matches_attrs and matches_text:
    return elem
```
One could argue that making an element for each node is time and resource consuming, but I did run a few stress tests to check whether that is a scalable solution, and it is! Feel free to perform tests yourself.

I ran `format.sh` and `lint.sh` and here is the result:
![image](https://github.com/user-attachments/assets/4eef53cd-7e47-42b7-9bdb-3fbedf968ca5)
I couldn't understand why `lint.sh` showed two errors while I only declared an empty list called tasks. I think that it can be ignored..?

I've put so much work into this pull request, I'd love to see it merged.
Let me know what you think about the changes I made

Regards,
Hamza

hamzaaitbrik commented:
Hey @stephanlensky, I did the changes you asked. Removed `find_element_by_tagname_attrs_text` and `find_elements_by_tagname_attrs_text` and added a function `_find_elements_by_tagname_attrs_text` that will be called using the `find` and `find_all` functions with the boolean argument `return_after_first_match`. I tested the code against many possibilities. I didn't run a stress test like before, but it should be about the same, or even better. Let me know what you think.

stephanlensky commented:
Hey Hamza, I'm sorry I actually did not mean to merge this. I checked out your branch locally and accidentally pushed it, now the main branch of the repo is slightly messed up. Apologies for the confusion, let me try to fix this and we may then need to open a new PR.

stephanlensky commented:
Okay, I fixed it, we just need to open a new PR.

Sorry again, I was trying to help you by merging latest main into your branch, but I accidentally did the opposite...
